src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -199,16 +188,48 @@ private JCTree opToTree(Op op) {
                 var val = arrayStoreOp.operands().get(1);
                 var index = arrayStoreOp.operands().get(2);
                 var as = treeMaker.Assign(
-                        treeMaker.Indexed((JCTree.JCExpression) valueToTree.get(array), (JCTree.JCExpression) valueToTree.get(index)),
-                        (JCTree.JCExpression) valueToTree.get(val)
+                        treeMaker.Indexed(exprTree(array), exprTree(index)), exprTree(val)
                 );
                 yield treeMaker.Exec(as);
                 // body builder are created but never passed when creating the op, why ?
             }
             default -> throw new IllegalStateException("Op -> JCTree not supported for :" + op.getClass().getName());
         };
-        valueToTree.put(op.result(), tree);
-        return tree;
+        if (tree instanceof JCTree.JCExpression expr) {
+            // introduce a local variable to hold the expr, to make sure an op's tree is inserted right away
+            // for some operations this is essential, e.g. to ensure the correct order of operations
+            Type type;
+            if (op instanceof CoreOp.ConstantOp cop && cop.value() == null) {
+                // if ConstantOp value is null, tree.type will be null_type
+                // if null_type is used to create a VarSymbol, an exception will be thrown
+                type = typeElementToType(cop.resultType());
+            } else {
+                type = tree.type;
+            }
+            var vs = new Symbol.VarSymbol(LocalVarFlags, names.fromString("_$" + c++), type, ms);
+            var varDef = treeMaker.VarDef(vs, expr);
+            map(op.result(), varDef);
+            return varDef;
+        } else {
+            map(op.result(), tree);
+            return tree;
+        }
+    }
+
+    private JCTree.JCExpression exprTree(Value v) {
```

Comment:
Why do we have nodes in the `valueToTree` map that are not variables?

Reply:
It seems to me that, with the exception of `return` we only deal with expressions? I see that we create some intermediate `Exec` nodes (e.g. for array access) but it seems to me that this indirection can be removed in this new code?

Reply:
It seems to me that, with the exception of `return` we only deal with expressions? I see that we create some intermediate `Exec` nodes (e.g. for array access) but it seems to me that this indirection can be removed in this new code?

Reply:
Can you clarify ?